nixos/tests/paperless.nix

```diff
@@ -2,65 +2,88 @@ import ./make-test-python.nix ({ lib, ... }: {
   name = "paperless";
   meta.maintainers = with lib.maintainers; [ erikarvstedt Flakebi ];
 
-  nodes.machine = { pkgs, ... }: {
-    environment.systemPackages = with pkgs; [ imagemagick jq ];
-    services.paperless = {
-      enable = true;
-      passwordFile = builtins.toFile "password" "admin";
+  nodes = let self = {
+    simple = { pkgs, ... }: {
+      environment.systemPackages = with pkgs; [ imagemagick jq ];
+      services.paperless = {
+        enable = true;
+        passwordFile = builtins.toFile "password" "admin";
+      };
     };
-  };
+    postgres = { config, pkgs, ... }: {
+      imports = [ self.simple ];
+      services.postgresql = {
+        enable = true;
+        ensureDatabases = [ "paperless" ];
+        ensureUsers = [
+          { name = config.services.paperless.user;
+            ensurePermissions = { "DATABASE \"paperless\"" = "ALL PRIVILEGES"; };
+          }
+        ];
+      };
+      services.paperless.extraConfig = {
+        PAPERLESS_DBHOST = "/run/postgresql";
+      };
+    };
```

Comment:
I introduced this change. `rec` is a nix antipattern that should be avoided in larger attrsets ([1](https://nix.dev/recipes/best-practices#recursive-attribute-set-rec), [2](https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293)). But I have no strong preference here.

Reply:
Note that nix.dev is an opinionated guide; there is no broader consensus on `rec` that I'm aware of. What I believe everyone can agree with though is that `rec` can be confusing when used in expressions that can be overridden such as mkDerivation but, while an alternative exists, it has performance drawbacks, so even there isn't a clear-cut reason to avoid rec. This isn't the case here however.

I personally think a rec is easier to read in many cases but I'm not totally against patterns like this; it just stood out as a bit weird to me.